install.sh

```diff
@@ -296,176 +293,309 @@ int main() {
-    # We combine these two things into one glorious all-inclusive file
-    # for the `curl` command. It is just as private as id_rsa.
-    cat var/sandcats/id_rsa var/sandcats/id_rsa.pub > var/sandcats/id_rsa.private_combined
+  # We look for sysvinit, as a convenient fallback. Note that this
+  # should work fine with Upstart (on e.g. Ubuntu 14.04), too.
+  if [ -e /etc/init.d ]; then
+    INIT_SYSTEM="sysvinit"
+    return
+  fi
 
-    # Set filesystem permissions, in case the files get copied into the
-    # wrong place later.
-    chmod 0600 var/sandcats/id_rsa var/sandcats/id_rsa.pub var/sandcats/id_rsa.private_combined
+  # If we got this far, and we couldn't figure out the init system
+  # in use, that's life.
+}
 
-    # Go to the start of the line, before the "..." that we left on
-    # the screen, allowing future echo statements to overwrite it.
-    echo -ne '\r'
+choose_install_mode() {
+  if [ "yes" = "$USE_DEFAULTS" ] ; then
+    CHOSEN_INSTALL_MODE=2  # dev server mode
   fi
 
-  # Having set up the keys, we run the function to register a name
-  # with Sandcats. This function handles tail-recursing itself until
-  # it succeeds and/or returning when the user expresses a desire to
-  # cancel the process.
-  register_sandcats_name
-fi
-
-# ========================================================================================
-# Write config
+  if [ -z "${CHOSEN_INSTALL_MODE:-}" ]; then
+    echo "Sandstorm makes it easy to run web apps on your own server. You can have:"
+    echo ""
+    echo "1. A full server with automatic setup (press enter to accept this default)"
+    echo "2. A development server, for writing apps."
+    echo ""
+    CHOSEN_INSTALL_MODE=$(prompt "How are you going to use this Sandstorm install?" "1")
+  fi
 
-writeConfig() {
-  while [ $# -gt 0 ]; do
-    eval echo "$1=\$$1"
-    shift
-  done
+  if [ "1" = "$CHOSEN_INSTALL_MODE" ] ; then
+    full_server_install
+  else
+    dev_server_install
+  fi
 }
 
-# TODO(someday): Ask what channel to use. Currently there is only one channel.
-CHANNEL=dev
+dev_server_install() {
+  # Use these settings for a dev-server-oriented install.
+  #
+  # Users will find themselves going through this flow if they
+  # manually choose a dev-server-related flow, but also if they pass
+  # -d on the command line. (The Vagrantfile and the test suite both
+  # use -d. The test suite runs install.sh with -d -u.)
+  #
+  # A "dev server install" must be run as root, unless you pass
+  # -u. That's because app development (aka spk dev) requires running
+  # as root, at the moment.
+
+  if [ "yes" = "$PREFER_ROOT" ] && [ "no" = "$CURRENTLY_UID_ZERO" ] ; then
+    # We are not root, but we would like to be root.
+    echo "If you want app developer mode for a Sandstorm install, you need root"
+    echo "due to limitations in the Linux kernel."
+    echo ""
+
+    echo "To set up Sandstorm, we will need to use sudo."
+    echo "Rest assured that Sandstorm itself won't run as root."
```

Comment:
This is debatable, in this case. The sandstorm bundle runner actually does run as root, but runs Mongo and Node as non-root.